lms/djangoapps/courseware/module_render.py

```diff
@@ -554,12 +554,14 @@ def handle_deprecated_progress_event(block, event):
             if requested_user_id != user.id:
                 log.warning("{} tried to submit a completion on behalf of {}".format(user, requested_user_id))
                 return
-            BlockCompletion.objects.submit_completion(
-                user=user,
-                course_key=course_id,
-                block_key=block.scope_ids.usage_id,
-                completion=1.0,
-            )
+
+            if not hasattr(block, 'has_custom_completion'):
```

Comment:
Should be `if not getattr(block, `has_custom_completion`, False):` otherwise this can't handle blocks with `has_custom_completion = False`

Reply:
Also, please add a comment to explain that if blocks explicitly declare support for the new completion API, we expect them to emit 'completion' events, and we ignore the deprecated 'progress' events in order to avoid duplicate work and support slightly different semantics.

Reply:
👍